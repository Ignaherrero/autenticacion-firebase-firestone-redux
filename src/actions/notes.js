import Swal from "sweetalert2";
import { db } from "../firebase/firebase-config";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import types from "../types/types";

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    const newNote = {
      title: "",
      body: "",
      date: new Date().getTime(),
    };
    const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);
    dispatch(activeNotes(docRef.id, newNote));
    dispatch(addNewNote(docRef.id, newNote));
  };
};

export const activeNotes = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note,
  },
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const setNotes = (note) => ({
  type: types.notesLoad,
  payload: note,
});

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    try {
      const { uid } = getState().auth;
      if (!note.url) delete note.url;
      if (!note.img) delete note.img;

      const noteToFirestone = { ...note };
      delete noteToFirestone.id;

      await db.doc(`/${uid}/journal/notes/${note.id}`).update(noteToFirestone);
      dispatch(refreshNote(note.id, note));
      Swal.fire("Save", note.title, "success");
    } catch (error) {
      console.log(error);
    }
  };
};

export const refreshNote = (id, note) => ({
  type: types.notesUpdate,
  payload: { id, note },
});

export const startUploadingImage = (file) => {
  return async (dispatch, getState) => {
    Swal.fire({
      title: "Uploading...",
      text: "Please wait...",
      allowOutsideClick: false,
      didOpen: () => {
        Swal.showLoading();
      },
    });
    const { active: activeNote } = getState().notes;

    const fileUrl = await fileUpload(file);
    Swal.close();
    dispatch(startSaveNote({ ...activeNote, url: fileUrl }));
    dispatch(activeNotes(activeNote.id, { ...activeNote, url: fileUrl }));
  };
};

export const startDeleteNote = (id) => {
  return async (dispatch, getState) => {
    try {
      const { uid } = getState().auth;
      await db.doc(`/${uid}/journal/notes/${id}`).delete();
      dispatch(deleteNote(id));
    } catch (error) {
      console.log(error);
    }
  };
};

export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: id,
});

export const noteLogout = () => ({
  type: types.notesLogoutCleaning,
  payload: { notes: [], active: null, date: null, body: "" },
});

export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: { id, ...note },
});
