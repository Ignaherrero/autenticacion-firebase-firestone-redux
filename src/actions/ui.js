import types from "../types/types";

export const setError = (error) => ({
  type: types.uiSelector,
  payload: error,
});

export const removeError = () => ({
  type: types.uiRemoveError,
});

export const startLogin = () => ({
  type: types.uiStartLoading,
});

export const finishLogin = () => ({
  type: types.uiFinishLoading,
});
