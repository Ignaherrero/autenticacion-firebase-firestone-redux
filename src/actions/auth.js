import Swal from "sweetalert2";
import { firebase, googleAuthProvider } from "../firebase/firebase-config";
import types from "../types/types";
import { noteLogout } from "./notes";
import { finishLogin, startLogin } from "./ui";

export const login = (uid, uname) => ({
  type: types.login,
  payload: { uid, uname },
});

export const logOut = () => ({
  type: types.logout,
});

export const starRegisterWithEmail = ({ email, password, name }) => {
  return (dispatch) => {
    dispatch(startLogin());
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        dispatch(login(user.uid, user.displayName));
      })
      .catch(({ message }) => {
        Swal.fire({
          icon: "error",
          title: "Oops...",
          text: message,
          footer: '<a href="">Why do I have this issue?</a>',
        });
      });
  };
};

export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {
    dispatch(startLogin());
    return firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLogin());
      })
      .catch(({ message }) => {
        Swal.fire({
          icon: "error",
          title: "Oops...",
          text: message,
          footer: '<a href="">Why do I have this issue?</a>',
        });
        dispatch(finishLogin());
      });
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    dispatch(startLogin());
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLogin());
      });
  };
};

export const startLogout = () => {
  return async (dispatch) => {
    await firebase.auth().signOut();
    dispatch(logOut());
    dispatch(noteLogout());
  };
};
