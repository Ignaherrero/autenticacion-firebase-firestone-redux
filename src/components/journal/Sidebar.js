import React from "react";
import { startLogout } from "../../actions/auth";
import { JournalEntries } from "./JournalEntries";
import { useDispatch, useSelector } from "react-redux";
import { startNewNote } from "../../actions/notes";

export const Sidebar = () => {
  const dispatch = useDispatch();
  const handleLogout = () => {
    dispatch(startLogout());
  };
  const { auth } = useSelector((state) => state);
  const { uname } = auth;
  const handleAddTask = () => {
    dispatch(startNewNote());
  };

  return (
    <aside className="journal__sidebar">
      <div className="journal__sidebar-navbar">
        <h3 className="mt-5">
          <i className="far fa-moon"></i>
          <span> {uname} </span>
        </h3>

        <button className="btn" onClick={handleLogout}>
          Logout
        </button>
      </div>

      <div className="journal__new-entry" onClick={handleAddTask}>
        <i className="far fa-calendar-plus fa-5x"></i>
        <p className="mt-5">New entry</p>
      </div>

      <JournalEntries />
    </aside>
  );
};
