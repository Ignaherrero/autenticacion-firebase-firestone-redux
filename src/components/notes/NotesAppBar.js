import moment from "moment";
import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { startSaveNote, startUploadingImage } from "../../actions/notes";

export const NotesAppBar = () => {
  const { active } = useSelector((state) => state.notes);
  const noteDate = moment(active.date);
  const dispatch = useDispatch();
  const handleSaveNote = () => {
    dispatch(startSaveNote(active));
  };
  const handlePictureClick = () => {
    document.querySelector("#fileSelector").click();
  };
  const handleChangeImage = (e) => {
    const file = e.target.files[0];
    if (file) {
      dispatch(startUploadingImage(file));
    }
  };
  // TODO quitar el estilo en linea
  return (
    <div className="notes__appbar">
      <span>{noteDate.format("L")}</span>
      <input
        name="file"
        id="fileSelector"
        type="file"
        style={{ display: "none" }}
        onChange={handleChangeImage}
      />
      <div>
        <button className="btn" onClick={handlePictureClick}>
          Picture
        </button>

        <button className="btn" onClick={handleSaveNote}>
          Save
        </button>
      </div>
    </div>
  );
};
