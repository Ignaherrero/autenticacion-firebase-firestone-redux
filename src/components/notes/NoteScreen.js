import React, { useEffect, useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import { activeNotes, startDeleteNote } from "../../actions/notes";
import { useForm } from "../../hooks/useForm";
import { NotesAppBar } from "./NotesAppBar";

export const NoteScreen = () => {
  const { active: note } = useSelector((state) => state.notes);
  const [values, handleInputChange, reset] = useForm(note);
  const { title, body } = values;
  const dispatch = useDispatch();
  let activeId = useRef(note.id);

  useEffect(() => {
    if (note.id !== activeId.current) {
      reset(note);
      activeId.current = note.id;
    }
  }, [note, reset]);

  useEffect(() => {
    dispatch(activeNotes(values.id, values));
  }, [dispatch, values]);

  const handleDeleteNote = (id) => {
    dispatch(startDeleteNote(id));
  };

  return (
    <div className="notes__main-content">
      <NotesAppBar />

      <div className="notes__content">
        <input
          type="text"
          placeholder="Some awesome title"
          className="notes__title-input"
          autoComplete="off"
          name="title"
          value={title}
          onChange={handleInputChange}
        />

        <textarea
          placeholder="What happened today"
          className="notes__textarea"
          name="body"
          value={body}
          onChange={handleInputChange}
        >
          {title}
        </textarea>

        <div className="notes__image">
          <img src={note.url} alt="imagen" />
        </div>
        <button
          className="btn btn-dangerous"
          onClick={() => handleDeleteNote(note.id)}
        >
          Delete
        </button>
      </div>
    </div>
  );
};
