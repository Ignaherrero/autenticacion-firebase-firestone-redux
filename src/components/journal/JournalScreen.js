import React from "react";
import { Sidebar } from "./Sidebar";
import { NoteScreen } from "../notes/NoteScreen";
import { NothingSelected } from "./NothingSelected";
import { useSelector } from "react-redux";

export const JournalScreen = () => {
  const { notes } = useSelector((state) => state);

  return (
    <div className="journal__main-content animate__animated animate__fadeIn animate__fast">
      <Sidebar />

      <main>{notes.active ? <NoteScreen /> : <NothingSelected />}</main>
    </div>
  );
};
