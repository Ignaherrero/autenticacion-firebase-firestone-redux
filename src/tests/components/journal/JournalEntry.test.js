import { mount } from "enzyme";
import React from "react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { MemoryRouter } from "react-router-dom";
import { JournalEntry } from "../../../components/journal/JournalEntry";
import { activeNotes } from "../../../actions/notes";

jest.mock("../../../actions/notes", () => ({
  activeNotes: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};

let store = mockStore(initState);
store.dispatch = jest.fn();

const note = {
  id: "123",
  title: "algo",
  body: "algo mas",
  url: "www.google.com",
  img: "http://www.imagenes.com/",
  date: new Date(),
};

const wrapper = mount(
  <MemoryRouter>
    <Provider store={store}>
      <JournalEntry {...note} />
    </Provider>
  </MemoryRouter>
);
describe("Pruebas en <JorunalEntry/>", () => {
  test("debe renderizar el componente correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("debe disparar la acciÃ³n activeNotes", () => {
    wrapper.find(".journal__entry").prop("onClick")();
    expect(activeNotes).toHaveBeenCalled();
    expect(store.dispatch).toHaveBeenCalledWith(activeNotes(note.id, note));
  });
});
