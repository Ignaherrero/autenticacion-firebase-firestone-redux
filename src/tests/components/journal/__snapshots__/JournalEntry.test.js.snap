// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <JorunalEntry/> debe renderizar el componente correctamente 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn animate__fast"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(www.google.com)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      algo
    </p>
    <p
      className="journal__entry-content"
    >
      algo mas
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Monday
    </span>
    <h4>
      12
    </h4>
  </div>
</div>
`;
